scripts/benchmark_db/benchmark_base.rb

```diff
@@ -0,0 +1,263 @@
+# frozen_string_literal: true
+
+# Mixin module for base benchmark class exec (and exec helper) commands.
+module ExecCommands
+  # Helper function used in calculation of number of epochs.
+  def get_last_epoch(benchmark)
+    epoch = nil
+    while epoch.nil?
+      # Epoch can be nil (e.g. if the client is in the "fetchting messages" stage).
+      epoch = benchmark.epoch_command
+      sleep 0.5
+    end
+    epoch
+  end
+
+  # Measures validation time for daily metrics.
+  def measure_online_validation(benchmark, pid, metrics)
+    Thread.new do
+      first_epoch = nil
+      loop do
+        start, first_epoch = benchmark.start_online_validation_command
+        if start
+          @logger.info 'Start measure'
+          break
+        end
+        sleep 0.1
+      end
+      sleep benchmark.online_validation_secs
+      last_epoch = get_last_epoch(benchmark)
+      metrics[:num_epochs] = last_epoch - first_epoch
+
+      @logger.info 'Stopping process...'
+      benchmark.stop_command(pid)
+    end
+  end
+
+  # Calls online validation function and runs monitor to measure memory usage.
+  def proc_monitor(pid, benchmark)
+    metrics = Concurrent::Hash.new
+    metrics[:rss] = []
+    metrics[:vsz] = []
+    measure_online_validation(benchmark, pid, metrics) if benchmark
+    handle = Thread.new do
+      loop do
+        sample_proc(pid, metrics)
+        sleep 0.5
+      # Loop will error during clean and build steps, so need to break until
+      # client is running.
+      rescue StandardError => _e
+        break
+      # Need to handle interrupt or process will continue on `ctrl-c`.
+      rescue Interrupt
+        stop_command(pid)
+        break
+      end
+    end
+    [handle, metrics]
+  end
+
+  # Helper function for measuring execution time; passes process ID to online
+  # validation and process monitor.
+  def exec_command_aux(command, metrics, benchmark)
+    Open3.popen2(*command) do |i, o, t|
+      pid = t.pid
+      i.close
+
+      handle, proc_metrics = proc_monitor(pid, benchmark)
+      o.each_line do |l|
+        print l
+      end
+
+      handle.join
+      metrics.merge!(proc_metrics)
+    end
+  end
+
+  # Measures execution time of command.
+  def exec_command(command, benchmark = nil)
+    @logger.info "$ #{command.join(' ')}"
+    return {} if @dry_run
+
+    metrics = Concurrent::Hash.new
+    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    exec_command_aux(command, metrics, benchmark)
+    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    metrics[:elapsed] = trunc_seconds(t1 - t0)
+    metrics
+  end
+end
+
+# Mixin module for base benchmark class build commands.
+module BuildCommands
+  # Helper function to build config file for `build_artefacts` function
+  def build_config_file
+    config = @config.deep_merge(default_config)
+    config_path = "#{data_dir}/#{@name}.toml"
+
+    toml_str = TomlRB.dump(config)
+    File.write(config_path, toml_str)
+  end
+  private :build_config_file
```

Comment:
You don't need to do it this way as far as I remember. You can just put `private` and every method below this keyword will be private (until it hits the end of the class/module or another relevant keyword)

Reply:
This is a carryover from the original script so I will have to defer to @elmattic to explain if there was a reason for doing it this way. If not, I can change it.

Reply:
@jdjaustin no reason other than being new to Ruby. Will address Hubert comment.